direct.go

```diff
@@ -208,44 +272,59 @@ func (d *direct) dialWith(in chan *Masquerade, network string) (net.Conn, bool,
 		// we expect.
-		if conn, err := d.dialServerWith(m); err != nil {
-			log.Tracef("Could not dial to %v, %v", m.IpAddress, err)
-			// Don't re-add this candidate if it's any certificate error, as that
-			// will just keep failing and will waste connections. We can't access the underlying
-			// error at this point so just look for "certificate" and "handshake".
-			if strings.Contains(err.Error(), "certificate") || strings.Contains(err.Error(), "handshake") {
-				log.Tracef("Not re-adding candidate that failed on error '%v'", err.Error())
-			} else {
-				log.Tracef("Unexpected error dialing, keeping masquerade: %v", err)
-				retryLater = append(retryLater, m)
-			}
-		} else {
-			log.Tracef("Got successful connection to: %v", m)
-			if err := d.headCheck(m); err != nil {
-				log.Tracef("Could not perform successful head request: %v", err)
-			} else {
-				m.LastVetted = time.Now()
-				// Requeue the working connection to masquerades
-				d.masquerades <- m
-				select {
-				case d.toCache <- m:
-					// ok
-				default:
-					// cache writing has fallen behind, drop masquerade
+		conn, retriable, err := d.doDial(m)
+		if err == nil {
+			log.Trace("Returning connection")
+			masqueradeGood := func(good bool) bool {
+				if good {
+					m.LastVetted = time.Now()
+					// Requeue the working connection to masquerades
+					d.masquerades <- m
+					select {
+					case d.toCache <- m:
+						// ok
+					default:
+						// cache writing has fallen behind, drop masquerade
+					}
+				} else {
+					go d.vetOneUntilGood()
 				}
-				idleTimeout := 70 * time.Second
-
-				log.Trace("Wrapping connecting in idletiming connection")
-				conn = idletiming.Conn(conn, idleTimeout, func() {
-					log.Tracef("Connection to %v idle for %v, closed", conn.RemoteAddr(), idleTimeout)
-				})
-				log.Trace("Returning connection")
-				return conn, true, nil
+
+				return good
 			}
+			return conn, masqueradeGood, true, err
+		} else if retriable {
+			retryLater = append(retryLater, m)
 		}
 	}
 }
 
+func (d *direct) doDial(m *Masquerade) (conn net.Conn, retriable bool, err error) {
+	conn, err = d.dialServerWith(m)
+	if err != nil {
+		log.Tracef("Could not dial to %v, %v", m.IpAddress, err)
+		// Don't re-add this candidate if it's any certificate error, as that
+		// will just keep failing and will waste connections. We can't access the underlying
+		// error at this point so just look for "certificate" and "handshake".
+		if strings.Contains(err.Error(), "certificate") || strings.Contains(err.Error(), "handshake") {
+			log.Tracef("Not re-adding candidate that failed on error '%v'", err.Error())
+			retriable = false
+		} else {
+			log.Tracef("Unexpected error dialing, keeping masquerade: %v", err)
+			retriable = true
+		}
+	} else {
+		log.Tracef("Got successful connection to: %v", m)
+		idleTimeout := 70 * time.Second
+
+		log.Trace("Wrapping connecting in idletiming connection")
+		conn = idletiming.Conn(conn, idleTimeout, func() {
```

Comment:
I don't think we can enable connection reuse, because the Transport does that based on the url of the HTTP request, which means that in effect it won't cycle through the different masquerades. If one of the masquerades goes bad, we'll potentially be in trouble.